flowapi/flowapi/permissions.py

```diff
@@ -0,0 +1,358 @@
+            ),
+        )
+    ] + [
+        "get_result:available_dates"
+    ]  # Special case for available dates
+
+
+def schema_to_scopes(schema: dict) -> Iterable[str]:
+    """
+    Constructs and yields query scopes of the form:
+    <action>:<query_kind>:<arg_name>:<arg_val>
+    where arg_val may be a query kind, or the name of an aggregation unit if applicable, and <action> is run or get_result.
+    Additionally yields the "get_result:available_dates" scope.
+
+    One scope is yielded for each viable query structure, so for queries which contain two child queries
+    five scopes are yielded. If that query has 3 possible aggregation units, then 13 scopes are yielded altogether.
+
+    Parameters
+    ----------
+    flowmachine_query_schemas : dict
+        Schema dict to turn into scopes list
+
+    Yields
+    ------
+    str
+        Scope strings
+
+    Examples
+    --------
+    >>>list(schema_to_scopes({"FlowmachineQuerySchema": {"oneOf": [{"$ref": "DUMMY"}]},"DUMMY": {"properties": {"query_kind": {"enum": ["dummy"]}}},},))
+    ["get_result:dummy", "run:dummy", "get_result:available_dates"],
+    """
+    return make_scopes(
+        queries=ResolvingParser(spec_string=dumps(schema)).specification["components"][
+            "schemas"
+        ]["FlowmachineQuerySchema"]
+    )
+
+
+def expand_scopes(*, scopes: List[str]) -> str:
+    """
+    Expand up a list of compact scopes to full scopes
+
+    Parameters
+    ----------
+    scopes : list of str
+        Compressed scopes to expand
+
+    Yields
+    ------
+    str
+        A scope string
+
+    """
+    for scope in scopes:
+        parts = scope.split(":")
+        ps = (x.split(",") for x in parts)
+        yield from (":".join(x) for x in product(*ps))
+
+
+def q_to_scope_atoms(
+    *, query: dict, argument_names_to_extract: List[str] = ["aggregation_unit"],
+) -> List[Union[str, List[str]]]:
+    """
+    Convert a query to scope 'atoms', each of which would be a contiguous
+    part of a scope string.
+
+    Parameters
+    ----------
+    query : dict
+        Query to atomise
+    argument_names_to_extract : list of str
+        Arguments to pull out
+
+    Returns
+    -------
+    list of str
+       Nested list of atoms
+    """
+    paths = []
+    if "query_kind" in query:
+        paths = paths + [query["query_kind"]]
+    for k, v in sorted(query.items()):
+        if k in argument_names_to_extract:
+            paths = paths + [k, v]
+        if isinstance(v, dict):
+            atoms = q_to_scope_atoms(query=v)
+            if len(atoms) > 0:
+                paths = paths + [k] + atoms
+        if isinstance(v, list):
+            child_paths = set()
+            for x in v:
+                if isinstance(x, dict):
+                    child_paths.add(tuple(q_to_scope_atoms(query=x)))
+            for child in child_paths:
+                if len(child) > 0:
+                    paths = paths + [k] + list(child)
```

Comment:
A lot of the issue I'm having here is that I've gone with (to facilitate rendering FlowAuth side) a canonical ordering for the parts of the scope string which is based on how many options each argument has. That gives you quite a nice ordering for navigating permissions, but, the information required to reconstruct that ordering obviously isn't available when receiving just a query.

Reply:
In this specific example though, I think the behaviour is actually what I'd want - while we can't actually express it in the schema, the list items here should be homogenous in terms of of method & type, otherwise you're gonna get a result that is gibberish.

Reply:
Does the ordering need to be done on how many options each argument has? Ordering alphabetically by argument name would also give a unique order, and that information is available when receiving just a query.

Reply:
No, one just needs to pick one and stick to it really. The reason to do the ordering by number of options is that it gives you a slightly nicer tree to navigate in flowauth when setting the permissions (in particular for joined spatial aggregate).

Reply:
Ah, I see. I suppose the ordering doesn't necessarily have to be the same in FlowAPI and FlowAuth - the process of expanding a squashed scope requires splitting it into its parts, so it wouldn't be too hard to reorder the parts at the same time. Then the number-of-options ordering could be used for rendering in FlowAuth (where we have the full set of possible scopes), and alphabetical ordering could be used to check user scopes in FlowAPI (where we're building a scope from a single query, so don't have information about the number of options).

Reply:
Yeah - I’m thinking that part of the issue here is I’m just trying to be a bit too lazy, and evade a presentation layer but at the cost of considerable pain elsewhere.